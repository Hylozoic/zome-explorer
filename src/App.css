.App {
  font-family: Verdana, Helvetica, Arial, sans-serif;
  padding: 50px;
} 

.dna-name {
  font-size: 20px;
  font-variant: small-caps;
  margin-bottom: 5px;
}

.dna-description {
  font-size: 14px;
  color: #555;
  margin-bottom: 15px;
}

.instance-id-label {
  font-size: 12px;
  margin-right: 5px;
}

.zome-list {
  border: 1px solid lightgray;
  margin: 25px 0px;
}

.zome-list-header {
  padding: 10px;
}

.zome {
  border-top: 1px solid lightgray;
  padding: 10px;
}

.zome-name {
  display: flex;
  flex-direction: row;  
}

.zome-description {
  margin-top: 5px;
  margin-bottom: 15px;
  font-size: 14px;
  color: #555;
}

.zome-functions {
  border-bottom: 1px solid lightgray;
  padding-left: 5px;
}

.zome-functions-header {
  margin-bottom: 5px;
}

.zome-function {
  display: flex;
  flex-direction: column;
  border-top: 1px solid lightgray;
  border-left: 1px solid lightgray;  
  border-right: 1px solid lightgray;    
  padding: 10px 10px 10px 10px;
}

.function-name {
  display: flex;
  flex-direction: row;
  margin-bottom: 10px;
}

.call-button {
  font-size: 20px;
  padding: 3px 0px;
}

.expand-button-icon {
  margin-left: 10px;
  margin-top: 3px;
}

/* utils */

.clickable-div {
  cursor: pointer;
  user-select: none;
}