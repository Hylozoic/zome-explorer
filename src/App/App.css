.app {
  font-family: Verdana, Helvetica, Arial, sans-serif;
  padding: 50px;
} 

.column-wrapper {
  display: flex;
}

.column {
  flex-grow: 1;
  flex-basis: 0;
}

.resizable-box {
  box-sizing: border-box;
  flex: 1 1 auto;
  min-width: 0;
  overflow-x: scroll;
  padding: 5px;
}

.separator {
  width: 32px;
  padding: 0;
  flex: 0 0 auto;
}

.separator::before {
  content: '';
  display: block;
  width: 2px;
  height: 100%;
  margin: 0 auto;
  background: lightgray;

}

.dna-name {
  font-size: 20px;
  font-variant: small-caps;
  margin-bottom: 5px;
}

.dna-description {
  font-size: 14px;
  color: #555;
  margin-bottom: 15px;
}

.input-label {
  font-size: 12px;
  margin-right: 5px;
}

.zome-list {
  border: 1px solid #C455A6;
  margin: 25px 0px;
}

.zome-list-header {
  padding: 10px;
  color: #65A9CE;
}

.zome {
  border-top: 1px solid #C455A6;
  padding: 10px;
}

.zome-name {
  display: flex;
  flex-direction: row;  
}

.zome-description {
  margin-top: 5px;
  margin-bottom: 15px;
  font-size: 14px;
  color: #555;
}

.zome-functions {
  border-bottom: 1px solid #C2F45D;
  margin-left: 5px;
}

.zome-functions-header {
  margin-bottom: 5px;
  color: #65A9CE;
}

.zome-functions-empty-state {
  font-size: 12px;  
  margin: 15px 0px;
  color: #555;
}

.zome-function {
  display: flex;
  flex-direction: column;
  border-top: 1px solid #C2F45D;
  border-left: 1px solid #C2F45D;  
  border-right: 1px solid #C2F45D;    
  padding: 10px 10px 10px 10px;
}

.zome-function-header {
  display: flex;
  flex-direction: column;
}

.function-name {
  display: flex;
  flex-direction: row;
}

.zome-function-signature {
  font-size: 12px;  
  margin: 10px 0px;
}

.zome-function-signature-label {
  margin: 5px 0px;
  color: #555;
}

.zome-function-signature-row {

}

.call-button {
  font-size: 20px;
  padding: 3px 0px;
}

.call-error {
  margin: 15px 0px;
  padding: 15px 20px;
  font-size: 14px; 
  border: 2px solid red;
}

.expand-button-icon {
  margin-left: 10px;
  margin-top: 3px;
}

.filter-row {
  margin-bottom: 25px;
  margin-top: 61px;
}

.call-record {
  border: 1px solid lightgray;
  margin: 5px 0px;
  padding: 0px 15px;
  overflow: scroll;
  max-width: 500px;
}

.call-record-path {
  margin: 10px 0px;
  font-size: 14px;  
}

.call-record-title {
  margin: 5px 0px;
  font-size: 12px;  
  color: #555;
}

/* utils */

.clickable-div {
  cursor: pointer;
  user-select: none;
}