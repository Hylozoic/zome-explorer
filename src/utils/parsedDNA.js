import rawDNA from '../dna/dna-metadata.json'
import { flow, pick, defaults, map, keys, filter } from 'lodash/fp'

export const META_SECTION_NAME = '__META__'

const { name, description } = rawDNA

export function parseZomes (dna) {
  return flow(
    keys,
    map(key => ({ name: key, ...dna.zomes[key] })),
    filter(zome => zome.name !== META_SECTION_NAME),
    map(pick(['name', 'description', 'fn_declarations'])),
    map(defaults({fn_declarations: [], description: ''}))
  )(dna.zomes)
}

const zomes = parseZomes(rawDNA)

export {
  rawDNA,
  name,
  description,
  zomes
}
