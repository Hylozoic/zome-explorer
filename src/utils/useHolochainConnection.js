import { useEffect, useRef } from 'react'
import { connect } from '@holochain/hc-web-client'

export default function useHolochainConnection (url) {
  var callZomeRef = useRef(() => {})

  useEffect(() => {
    var connection
    async function connectToConductor () {
      connection = await connect({url, wsClient: {max_reconnects: 0, reconnect_interval: 2500}})
      callZomeRef.current = connection.callZome
    }
    connectToConductor()
    return () => {
      if (connection && connection.close) {
        connection.close()
      }
    }
  })

  return callZomeRef
}