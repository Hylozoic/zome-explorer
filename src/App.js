import React, { useState } from 'react'
import './App.css'
import useHolochainConnection from './utils/useHolochainConnection'
import { name, description, zomes } from './utils/parsedDNA'
import ZomeList from './ZomeList';

export const submitCall = (callZome, { instanceId, zome, funcName, params }) => {
  return callZome(instanceId, zome, funcName)(JSON.parse(params))
}

function App() {
  const callZomeRef = useHolochainConnection()
  const [instanceId, setInstanceId] = useState('hylo-chat')

  const callZome = (zome, funcName) => callZomeRef.current(instanceId, zome, funcName)

  return (
    <div className="App">
      <div className='dna-name'>{name}</div>
      <div className='dna-description'>{description}</div>      
      <div>
        <label className='instance-id-label'>Instance Id</label>
        <input value={instanceId} onChange={({ target: { value }}) => setInstanceId(value)} />
      </div>
      <ZomeList callZome={callZome} zomes={zomes} />
    </div>
  )
}

export default App;
