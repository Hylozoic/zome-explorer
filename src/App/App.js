import React, { useState, useEffect } from 'react'
import './App.css'
import useHolochainConnection from '@hylozoic/react-holochain-hooks'
import { name, description, zomes } from '../utils/parsedDNA'
import ZomeList from '../ZomeList'
import Results from '../Results'

export default function App() {
  const [ready, setReady] = useState(false)
  const { callZomeRef } = useHolochainConnection(process.env.REACT_APP_ZOME_WEBSOCKET_URL, {}, setReady)
  const [instanceId, setInstanceId] = useState(sessionStorage.getItem('instanceId') || "")
  const storedHistory = sessionStorage.getItem('history')    
  const [history, setHistory] = useState((storedHistory && JSON.parse(storedHistory)) || [])  

  const setInstanceIdSession = value => {
    sessionStorage.setItem('instanceId', value)
    setInstanceId(value)
  } 

  useEffect(() => {
    document.title = `Zome Explorer - ${name}`
  })

  const callZome = zomeName => funcName => async params => {
    const result = await callZomeRef.current(instanceId, zomeName, funcName)(params)
    const callRecord = {
      path: `${instanceId}/${zomeName}/${funcName}`,
      params,
      result
    }
    const newHistory = [callRecord].concat(history)
    setHistory(newHistory)
    sessionStorage.setItem('history', JSON.stringify(newHistory))
    return callRecord
  }

  return (
    <div className='app'>
      <div className='column-wrapper'>
        <div className='column'>
          <div>
            <div className='dna-name'>{name}</div>
            <div className='dna-description'>{description}</div>      
            <div className='instance-input'>
              <label className='input-label'>Instance Id</label>
              <input value={instanceId} onChange={({ target: { value }}) => setInstanceIdSession(value)} />
            </div>
            <div className='status'>
              Status:&nbsp;
              {ready && <span className='connected'>Connected</span>}
              {!ready && <span className='disconnected'>Disconnected</span>}              
            </div>
            <ZomeList callZome={callZome} zomes={zomes} />
          </div>
        </div>
        <div className='separator' />
        <div className='column'>
          <Results history={history} />
        </div>
      </div>
    </div>
  )
}
