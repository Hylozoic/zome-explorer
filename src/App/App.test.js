import App from './App'
import { shallow } from 'enzyme'
import React from 'react'
import '../utils/parsedDNA'

jest.mock('../utils/parsedDNA', () => ({
  name: 'the dna name',
  description: 'any old dna',
  zomes: [
    {
      name: 'Zome One'
    },
    {
      name: 'Zome Two'
    }
  ]
}))

it('matches the last snapshot', () => {
  const wrapper = shallow(<App />)
  expect(wrapper).toMatchSnapshot()
})
