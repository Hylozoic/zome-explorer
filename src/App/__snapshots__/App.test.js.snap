// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the last snapshot 1`] = `
<div
  className="app"
>
  <div
    className="column-wrapper"
  >
    <div
      className="column"
    >
      <div>
        <div
          className="dna-name"
        >
          the dna name
        </div>
        <div
          className="dna-description"
        >
          any old dna
        </div>
        <div>
          <label
            className="input-label"
          >
            Instance Id
          </label>
          <input
            onChange={[Function]}
            value={null}
          />
        </div>
        <ZomeList
          callZome={[Function]}
          zomes={
            Array [
              Object {
                "name": "Zome One",
              },
              Object {
                "name": "Zome Two",
              },
            ]
          }
        />
      </div>
    </div>
    <div
      className="separator"
    />
    <div
      className="column"
    >
      <Results
        history={Array []}
      />
    </div>
  </div>
</div>
`;
